flink-yarn/src/main/java/org/apache/flink/yarn/WorkerSpecContainerResourceAdapter.java

```diff
@@ -71,52 +72,65 @@
 	}
 
 	Optional<Resource> tryComputeContainerResource(final WorkerResourceSpec workerResourceSpec) {
-		return Optional.ofNullable(workerSpecToContainerResource.computeIfAbsent(
+		final InternalContainerResource internalContainerResource = workerSpecToContainerResource.computeIfAbsent(
 			Preconditions.checkNotNull(workerResourceSpec),
-			this::createAndMapContainerResource));
+			this::createAndMapContainerResource);
+		if (internalContainerResource != null) {
+			return Optional.of(internalContainerResource.toResource());
+		} else {
+			return Optional.empty();
+		}
 	}
 
 	Set<WorkerResourceSpec> getWorkerSpecs(final Resource containerResource, final MatchingStrategy matchingStrategy) {
-		return getEquivalentContainerResource(containerResource, matchingStrategy).stream()
+		final InternalContainerResource internalContainerResource = new InternalContainerResource(containerResource);
+		return getEquivalentInternalContainerResource(internalContainerResource, matchingStrategy).stream()
 			.flatMap(resource -> containerResourceToWorkerSpecs.getOrDefault(resource, Collections.emptySet()).stream())
 			.collect(Collectors.toSet());
 	}
 
 	Set<Resource> getEquivalentContainerResource(final Resource containerResource, final MatchingStrategy matchingStrategy) {
+		final InternalContainerResource internalContainerResource = new InternalContainerResource(containerResource);
+		return getEquivalentInternalContainerResource(internalContainerResource, matchingStrategy).stream()
+			.map(InternalContainerResource::toResource)
+			.collect(Collectors.toSet());
+	}
+
+	private Set<InternalContainerResource> getEquivalentInternalContainerResource(final InternalContainerResource internalContainerResource, final MatchingStrategy matchingStrategy) {
```

Comment:
I would only keep the `Set<Resource> getEquivalentContainerResource(final Resource containerResource, final MatchingStrategy matchingStrategy)`. To avoid duplication code `final InternalContainerResource internalContainerResource = new InternalContainerResource(containerResource);`

Reply:
I see your point. But I think the benefit for the current approach is that, we make it explicit that for all methods visible from outside the first thing we do is to convert `Resource` into `InternalContainerResource`, and the last thing we do before returning is to covert it back.